cheroot/wsgi.py

```diff
@@ -254,6 +255,9 @@ def get_environ(self):
             # AF_UNIX. This isn't really allowed by WSGI, which doesn't
             # address unix domain sockets. But it's better than nothing.
             env['SERVER_PORT'] = ''
+            uid = self.req.conn.get_peer_uid()
+            if uid is not None:
+                env['_REMOTE_UID'] = str(uid)
```

Comment:
hmm.. cherrypy does not really have access to HTTPConnection objects, so I don't see how that would go back to the socket to do SO_PEERCRED. This is how it looks like (obviously I don't have intimate knowledge of these parts, I just did a quick browse of the related source code):

 * cherrypy instantiates a cheroot.wsgi.Server instance (in ServerAdapter). wsgi.Server is derived from the "plain" HTTP server, cheroot.server.HTTPServer
 * as a connection arrives, a HTTPConnection is instantiated to encapsulate the connection specific socket
 * we parse the incoming request, and a cheroot specific HTTPRequest is created, to encapsulate the low-level HTTP details. Here we still have a reference to the connection/server objects.
 * then a cheroot.wsgi.Gateway is created, one that takes the low-level HTTPRequest, and 
populates a WSGI environment, and passes it to the the WSGI application via a __call__ method. 
 * The environment is the only input to the WSGI app, even headers are translated into and from environment name-value pairs. (with ugly translations like Accept-Language becoming HTTP_ACCEPT_LANGUAGE)

There seems to be an intention to decouple the HTTP specific pieces and the app side, and the means to exchange information is the environment itself. It is thus not possible to go back from the app to HTTPConnection & HTTPRequest where the information is available.

The Native interface is one abstraction short (e.g. environment is not created and HTTPRequest is directly translated to cherrypy request objects), but that is not the default how cherrypy operates. That interface does not support authentication (e.g. request.login is not populated at all).

I could inject the socket into the environment (with a key like "wsgi.socket" to match other attributes), but my opinion is that this would be ugly and would derail the entire idea to decouple apps from HTTP specifics.

Short of adding this ugly hack to environment, we need to do the authentication on the cheroot side and pass the information via environment variables.

Still, REMOTE_USER looks to be the best solution, as that would work out of the box with cherrypy (e.g. request.login is filled automatically).

I could make the username resolution stuff optional by adding an additional constructor parameter to cheroot.wsgi.Server that could be driven by a cherrypy config variable (e.g. server.socket_file_auth True/False).

I can also amortize the performance impact by caching the username lookup, because in most cases, it would only be a _very_ limited number of users that would use the interface.

I can see these courses of action at this point if we are to use REMOTE_USER:
1) additional parameter w/o caching
2) additional parameter w/ caching
3) no additional parameter w/ caching

OR we can use X_REMOTE_UID, GID and PID as you suggested, in which case username resolution is not needed at all, so performance impact should be small. But in that case cherrypy.request.login will not be set automatically.

What do you think?